src/metadata/metadata_passive_io.c

```diff
@@ -0,0 +1,377 @@
+		ocf_cache_line_t cline)
+{
+	ocf_part_id_t lru_list;
+	struct ocf_lru_list *list;
+
+	lru_list = (cline % OCF_NUM_LRU_LISTS);
+	list = ocf_lru_get_list(&cache->free, lru_list, true);
+
+	OCF_METADATA_LRU_WR_LOCK(cline);
+	ocf_lru_remove_locked(cache, list, cline);
+	OCF_METADATA_LRU_WR_UNLOCK(cline);
+}
+
+static inline void remove_from_default(ocf_cache_t cache,
+		ocf_cache_line_t cline)
+{
+	ocf_part_id_t part_id = PARTITION_DEFAULT;
+	ocf_part_id_t lru_list;
+	struct ocf_lru_list *list;
+
+	lru_list = (cline % OCF_NUM_LRU_LISTS);
+	list = ocf_lru_get_list(&cache->user_parts[part_id].part, lru_list, false);
+
+	OCF_METADATA_LRU_WR_LOCK(cline);
+	ocf_lru_remove_locked(cache, list, cline);
+	OCF_METADATA_LRU_WR_UNLOCK(cline);
+
+	env_atomic_dec(&cache->user_parts[part_id].part.runtime->curr_size);
+
+}
+
+static void handle_previously_invalid(ocf_cache_t cache,
+		ocf_cache_line_t cline)
+{
+	ocf_core_id_t core_id;
+	uint64_t core_line;
+	uint32_t lock_idx = ocf_metadata_concurrency_next_idx(cache->mngt_queue);
+
+	ocf_metadata_get_core_info(cache, cline, &core_id, &core_line);
+
+	if (metadata_test_dirty(cache, cline) && core_id < OCF_CORE_MAX) {
+		ENV_BUG_ON(!metadata_test_valid_any(cache, cline));
+		/* Moving cline from the freelist to the default partitioin */
+		remove_from_freelist(cache, cline);
+
+		ocf_hb_cline_prot_lock_wr(&cache->metadata.lock, lock_idx, core_id,
+				core_line);
+		OCF_METADATA_LRU_WR_LOCK(cline);
+		ocf_cline_rebuild_metadata(cache, core_id, core_line, cline);
+		OCF_METADATA_LRU_WR_UNLOCK(cline);
+		ocf_hb_cline_prot_unlock_wr(&cache->metadata.lock, lock_idx, core_id,
+				core_line);
+		ocf_cleaning_init_cache_block(cache, cline);
+		ocf_cleaning_set_hot_cache_line(cache, cline);
+
+	} else {
+		/* Cline stays on the freelist*/
+
+		/* To prevent random values in the metadata fill it with the defaults */
+		metadata_init_status_bits(cache, cline);
+		ocf_metadata_set_core_info(cache, cline, OCF_CORE_MAX, ULLONG_MAX);
+	}
+}
+
+static void handle_previously_valid(ocf_cache_t cache,
+		ocf_cache_line_t cline)
+{
+	ocf_core_id_t core_id;
+	uint64_t core_line;
+	uint32_t lock_idx = ocf_metadata_concurrency_next_idx(cache->mngt_queue);
+
+	ocf_metadata_get_core_info(cache, cline, &core_id, &core_line);
+
+	if (metadata_test_dirty(cache, cline) && core_id < OCF_CORE_MAX) {
+		/* Cline stays on the default partition*/
+		ENV_BUG_ON(!metadata_test_valid_any(cache, cline));
+
+		remove_from_default(cache, cline);
+
+		ocf_hb_cline_prot_lock_wr(&cache->metadata.lock, lock_idx, core_id,
+				core_line);
+		OCF_METADATA_LRU_WR_LOCK(cline);
+		ocf_cline_rebuild_metadata(cache, core_id, core_line, cline);
+		OCF_METADATA_LRU_WR_UNLOCK(cline);
+		ocf_hb_cline_prot_unlock_wr(&cache->metadata.lock, lock_idx, core_id,
+				core_line);
+		ocf_cleaning_set_hot_cache_line(cache, cline);
+
+	} else {
+		/* Moving cline from the default partition to the freelist */
+		ocf_cleaning_purge_cache_block(cache, cline);
```

Comment:
ocf_cleaning_purge_cache_block attempts to increment cache->cleaner.refcnt - just double checking to make sure this isn't frozen in standby

Reply:
The counter is frozen only to switch active cleaning policy which is a disabled operation in standby mode